src/dune_rules/pkg_rules.ml

```diff
@@ -259,11 +273,251 @@ module Pkg = struct
+                 ]
+             in
+             let prefix = Path.outside_build_dir @@ installation_prefix t in
+             (* Append some commands to the install command that copy
+                the artifacts to their final installation directory. *)
+             Dune_lang.Action.Progn
+               [ action
+               ; Dune_lang.Action.Run
+                   [ Slang.text "mkdir"
+                   ; Slang.text "-p"
+                   ; Slang.text @@ Path.to_string @@ Path.parent_exn prefix
+                   ]
+               ; Dune_lang.Action.Run
+                   [ Slang.text "mv"
+                   ; (* Prevents mv from replacing the destination if it
+                        already exists. This can happen if two dune
+                        instances race to install the toolchain. Note
+                        that -n is not posix but it is supported by gnu
+                        coreutils and by the default mv command on
+                        macos, but not openbsd. *)
+                     Slang.text "-n"
+                   ; Slang.text
+                       (Path.to_string
+                        @@ installation_prefix_within_tmp_install_dir t tmp_install_dir)
+                   ; Slang.text @@ Path.to_string @@ Path.parent_exn prefix
+                   ]
+               ]
+           | _ ->
+             (* The install command is something other than `make install`, so don't
+                attempt to modify. *)
+             action)
+        | _ ->
+          (* Not a "run" action, so don't attempt to modify. *)
+          action)
+    ;;
+
+    (* A [Value_list_env.t] where the PATH variable is set to the toolchain's bin dir. *)
+    let env t =
+      let bin_dir = bin_dir t |> Path.outside_build_dir in
+      Value_list_env.add_path Value_list_env.empty Env_path.var bin_dir
+    ;;
+
+    (* An [Ocaml_toolchain.t] based on the toolchain installation. *)
+    let ocaml_toolchain t context =
+      let env = env t |> Value_list_env.to_env in
+      let bin_dir = bin_dir t in
+      let which prog =
+        let path = Path.Outside_build_dir.relative bin_dir prog in
+        let+ exists = Fs_memo.file_exists path in
+        if exists then Some (Path.outside_build_dir path) else None
+      in
+      let get_ocaml_tool ~dir:_ prog = which prog in
+      Ocaml_toolchain.make context ~which ~env ~get_ocaml_tool
+    ;;
+
+    (* An [Install_cookie.t] with a [bin] section populated with all
+       the executable files from the bin directory of the toolchain
+       installation. *)
+    let install_cookie t =
+      let bin_dir = bin_dir t in
+      Fs_memo.dir_contents bin_dir
+      >>| function
+      | Error _ -> { Install_cookie.files = Section.Map.empty; variables = [] }
+      | Ok files ->
+        let bin_paths =
+          Fs_cache.Dir_contents.to_list files
+          |> List.filter_map ~f:(fun (filename, kind) ->
+            match kind with
```

Comment:
Just in case there are other types of file in the bin directory (e.g. directories or devices), it makes sure we don't resolve `which` queries to those files.

Reply:
But why would there be such files? Does building the compiler create such entries?

Reply:
There shouldn't be such files when building the compiler. It's probably safe to remove this check.